py-polars/polars/internals/dataframe/groupby.py

```diff
@@ -90,90 +83,32 @@ def __init__(
         self.by = by
         self.maintain_order = maintain_order
 
-    def __iter__(self) -> Iterable[Any]:
-        groups_df = self._groups()
-        groups = groups_df["groups"]
+    def __iter__(self) -> GroupBy[DF]:
         df = self._dataframe_class._from_pydf(self._df)
-        for i in range(groups_df.height):
-            yield df[groups[i]]
-
-    def _select(self, columns: str | list[str]) -> GBSelection[DF]:  # pragma: no cover
-        """
-        Select the columns that will be aggregated.
+        by = [self.by] if isinstance(self.by, str) else self.by
 
-        Parameters
-        ----------
-        columns
-            One or multiple columns.
-
-        """
-        warnings.warn(
-            "accessing GroupBy by index is deprecated, consider using the `.agg`"
-            " method",
-            DeprecationWarning,
-        )
-        if isinstance(columns, str):
-            columns = [columns]
-        return GBSelection(
-            self._df,
-            self.by,
-            columns,
-            dataframe_class=self._dataframe_class,
-        )
-
-    def _select_all(self) -> GBSelection[DF]:
-        """Select all columns for aggregation."""
-        return GBSelection(
-            self._df,
-            self.by,
-            None,
-            dataframe_class=self._dataframe_class,
-        )
-
-    def _groups(self) -> DF:  # pragma: no cover
-        """
-        Get keys and group indices for each group in the groupby.
-
-        Returns
-        -------
-        DataFrame
-            A DataFrame with:
+        # Aggregate groups for any single column that is not specified as 'by'
+        non_by_cols = [c for c in df.columns if c not in by]
```

Comment:
Maybe use `pl.exlude(by)/pl.all().exclude(by)` here?

Reply:
Hm, not sure how that would help in this case.

Basically, I want to call `agg_groups` on a single column that is not among the `by` columns (calling it on `pl.all().exclude(by)` seems wasteful), and I want to have an alternate path if there is no such column.

I added a small performance improvement that avoids iterating over all columns.